File: test-data/load-test-data.js
const fs = require('fs');
const path = require('path');

require('dotenv').config({ path: path.join(__dirname, '..', '.env') });

const TEST_DATA_DIR = __dirname;
const TEST_DATA_PATH = path.join(TEST_DATA_DIR, 'test-data.json');
const TEST_DATA_EXAMPLE_PATH = path.join(TEST_DATA_DIR, 'test-data.example.json');

const REQUIRED_FIELDS = [
  { path: 'domain', env: 'BASE_URL', label: 'BASE_URL' },
  { path: 'test_users.A.id', env: 'TEST_USER_A_ID', label: 'TEST_USER_A_ID' },
  { path: 'test_users.A.password', env: 'TEST_USER_A_PASSWORD', label: 'TEST_USER_A_PASSWORD' },
  { path: 'test_users.B.password', env: 'TEST_USER_B_PASSWORD', label: 'TEST_USER_B_PASSWORD' },
  { path: 'test_users.B.name', env: 'TEST_USER_B_NAME', label: 'TEST_USER_B_NAME' },
  { path: 'test_users.B.initial_password', env: 'TEST_USER_B_INITIAL_PASSWORD', label: 'TEST_USER_B_INITIAL_PASSWORD' },
];

function getNestedValue(obj, keyPath) {
  return keyPath.split('.').reduce((current, key) => current?.[key], obj);
}

function setNestedValue(obj, keyPath, value) {
  const keys = keyPath.split('.');
  let current = obj;
  for (let i = 0; i < keys.length - 1; i++) {
    if (!current[keys[i]]) {
      current[keys[i]] = {};
    }
    current = current[keys[i]];
  }
  current[keys[keys.length - 1]] = value;
}

function applyEnvOverrides(testData) {
  const result = JSON.parse(JSON.stringify(testData));

  if (process.env.BASE_URL) {
    result.domain = process.env.BASE_URL;
  }
  if (process.env.TEST_USER_A_ID) {
    result.test_users.A.id = process.env.TEST_USER_A_ID;
  }
  if (process.env.TEST_USER_A_PASSWORD) {
    result.test_users.A.password = process.env.TEST_USER_A_PASSWORD;
  }
  if (process.env.TEST_USER_B_ID) {
    result.test_users.B.id = process.env.TEST_USER_B_ID;
  }
  if (process.env.TEST_USER_B_PASSWORD) {
    result.test_users.B.password = process.env.TEST_USER_B_PASSWORD;
  }
  if (process.env.TEST_USER_B_NAME) {
    result.test_users.B.name = process.env.TEST_USER_B_NAME;
  }
  if (process.env.TEST_USER_B_INITIAL_PASSWORD) {
    result.test_users.B.initial_password = process.env.TEST_USER_B_INITIAL_PASSWORD;
  }

  return result;
}

function validateTestData(testData) {
  const missing = REQUIRED_FIELDS
    .filter(({ path: fieldPath }) => !getNestedValue(testData, fieldPath))
    .map(({ label }) => label);

  if (missing.length > 0) {
    throw new Error(
      `필수 테스트 설정이 없습니다: ${missing.join(', ')}\n` +
      '`.env` 파일을 생성하세요: cp .env.example .env'
    );
  }
}

function resolveTestDataPath() {
  if (fs.existsSync(TEST_DATA_PATH)) {
    return TEST_DATA_PATH;
  }
  if (fs.existsSync(TEST_DATA_EXAMPLE_PATH)) {
    return TEST_DATA_EXAMPLE_PATH;
  }
  throw new Error(
    'test-data/test-data.json 또는 test-data/test-data.example.json 파일이 필요합니다.'
  );
}

function loadTestData({ validate = true } = {}) {
  const filePath = resolveTestDataPath();
  const testData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
  const merged = applyEnvOverrides(testData);

  if (validate) {
    validateTestData(merged);
  }

  return merged;
}

function getTestDataFilePath() {
  return TEST_DATA_PATH;
}

function ensureTestDataFile() {
  if (!fs.existsSync(TEST_DATA_PATH) && fs.existsSync(TEST_DATA_EXAMPLE_PATH)) {
    fs.copyFileSync(TEST_DATA_EXAMPLE_PATH, TEST_DATA_PATH);
  }
  return TEST_DATA_PATH;
}

module.exports = {
  loadTestData,
  applyEnvOverrides,
  getTestDataFilePath,
  ensureTestDataFile,
  TEST_DATA_PATH,
  TEST_DATA_EXAMPLE_PATH,
};

File: scripts/load_test_data.py
import json
import os
from pathlib import Path

from dotenv import load_dotenv

PROJECT_ROOT = Path(__file__).resolve().parent.parent
TEST_DATA_DIR = PROJECT_ROOT / 'test-data'
TEST_DATA_PATH = TEST_DATA_DIR / 'test-data.json'
TEST_DATA_EXAMPLE_PATH = TEST_DATA_DIR / 'test-data.example.json'

REQUIRED_FIELDS = [
    ('domain', 'BASE_URL', 'BASE_URL'),
    ('test_users.A.id', 'TEST_USER_A_ID', 'TEST_USER_A_ID'),
    ('test_users.A.password', 'TEST_USER_A_PASSWORD', 'TEST_USER_A_PASSWORD'),
    ('test_users.B.password', 'TEST_USER_B_PASSWORD', 'TEST_USER_B_PASSWORD'),
    ('test_users.B.name', 'TEST_USER_B_NAME', 'TEST_USER_B_NAME'),
    ('test_users.B.initial_password', 'TEST_USER_B_INITIAL_PASSWORD', 'TEST_USER_B_INITIAL_PASSWORD'),
]


def _get_nested_value(data, key_path):
    current = data
    for key in key_path.split('.'):
        if not isinstance(current, dict):
            return None
        current = current.get(key)
    return current


def _set_nested_value(data, key_path, value):
    keys = key_path.split('.')
    current = data
    for key in keys[:-1]:
        if key not in current:
            current[key] = {}
        current = current[key]
    current[keys[-1]] = value


def apply_env_overrides(test_data):
    result = json.loads(json.dumps(test_data))

    env_mappings = {
        'domain': 'BASE_URL',
        'test_users.A.id': 'TEST_USER_A_ID',
        'test_users.A.password': 'TEST_USER_A_PASSWORD',
        'test_users.B.id': 'TEST_USER_B_ID',
        'test_users.B.password': 'TEST_USER_B_PASSWORD',
        'test_users.B.name': 'TEST_USER_B_NAME',
        'test_users.B.initial_password': 'TEST_USER_B_INITIAL_PASSWORD',
    }

    for field_path, env_name in env_mappings.items():
        value = os.environ.get(env_name)
        if value:
            _set_nested_value(result, field_path, value)

    return result


def validate_test_data(test_data):
    missing = [
        label
        for field_path, _, label in REQUIRED_FIELDS
        if not _get_nested_value(test_data, field_path)
    ]

    if missing:
        raise ValueError(
            f"필수 테스트 설정이 없습니다: {', '.join(missing)}\n"
            "`.env` 파일을 생성하세요: cp .env.example .env"
        )


def resolve_test_data_path():
    if TEST_DATA_PATH.exists():
        return TEST_DATA_PATH
    if TEST_DATA_EXAMPLE_PATH.exists():
        return TEST_DATA_EXAMPLE_PATH
    raise FileNotFoundError(
        'test-data/test-data.json 또는 test-data/test-data.example.json 파일이 필요합니다.'
    )


def ensure_test_data_file():
    if not TEST_DATA_PATH.exists() and TEST_DATA_EXAMPLE_PATH.exists():
        TEST_DATA_PATH.write_text(
            TEST_DATA_EXAMPLE_PATH.read_text(encoding='utf-8'),
            encoding='utf-8',
        )
    return TEST_DATA_PATH


def load_test_data(validate=True):
    load_dotenv(PROJECT_ROOT / '.env')

    file_path = resolve_test_data_path()
    with open(file_path, 'r', encoding='utf-8') as f:
        test_data = json.load(f)

    merged = apply_env_overrides(test_data)

    if validate:
        validate_test_data(merged)

    return merged
